<H2>Control structure & function exercises

<h1 style="color:yellow;font-size:12px;">1．The Body Mass Index (BMI) is an internationally accepted standard for measuring the degree of obesity and health of a person. The formula is BMI = the weight/height (kg/m2). The adult BMI value in China is defined as:

Too thin: < 18.5 

Normal: 18.5-23.9 

Overweight: 24-27.9 

Fat: >= 28 

Please enter the weight and height, and output the corresponding BMI value and weight obesity judgment result. 

[input sample] 

60,1.6 

[output sample] 

Your BMI is 23.4 

normal  

[tips] 

The weight and height can be got by the input statement of "weight, height = eval(input())".

  [Reference code]

In [ ]:
weight, height = eval(input())
 
bmi = weight / height ** 2
print("Your BMI is {0:.1f}".format(bmi))
if bmi < 18.5:
      print("too thin")
elif bmi < 24:
      print("normal")
elif bmi < 27.9:
      print("overweight")
else:
      print("fat")

2. Convert the Fahrenheit temperature to the Celsius temperature according to the formula  C= 5/9×(F-32). Program and output a comparison table of the Fahrenheit(0~300) and the corresponding Celsius temperature(output every 20 degrees).

In [13]:
for f in range(0, 301, 20):
    c = 5 / 9 * (f - 32)
    print("{} f = {:.0f} c".format(f, c),end=',')

0 f = -18 c,20 f = -7 c,40 f = 4 c,60 f = 16 c,80 f = 27 c,100 f = 38 c,120 f = 49 c,140 f = 60 c,160 f = 71 c,180 f = 82 c,200 f = 93 c,220 f = 104 c,240 f = 116 c,260 f = 127 c,280 f = 138 c,300 f = 149 c,

3. Kakutani is a famous scholar in Japan. He proposed a conjecture: for a positive integer n, divide by 2 for even numbers, multiply by 3 plus 1 for odd numbers, and get a new one. Continue to calculate according to the previous two rules, the result obtained after several times must be 1. Enter any positive integer and output the calculation process

In [7]:
n = int(input())
while n != 1:
    if n % 2 == 0:
        print("{}/2={}".format(n, n//2))
        n //= 2                       
    else:
        print("{}*3+1={}".format(n, 3*n+1))
        n = 3*n+1

5*3+1=16
16/2=8
8/2=4
4/2=2
2/2=1


Enter n, use the recursion method (for example, derive the latter term from the relationship between the preceding items, it is a one-loop question) and caculate the sum of 1+2!+3!+...+n! and output the result.


In [9]:
n = int(input())
s = term = 1
 
for i in range(2, n+1):
    term *= i
    s += term
print(s)

4037913


5. Program to solve how many non-repetitive three-digit numbers can be composed of these four numbers of 1-4, and output these numbers in ascending order.

In [12]:
for i in range(1, 5):
    for j in range(1, 5):
        for k in range(1, 5):
            if i != k and i != j and j != k:
                print(100*i+10*j+k, end=',')



123,124,132,134,142,143,213,214,231,234,241,243,312,314,321,324,341,342,412,413,421,423,431,432,

6. Verify the proposition: If a three-digit integer is a multiple of 37, then the two other rotated three-digit numbers of the number are also a multiples of 37 (for example, 123 to 231 and 312).



In [21]:
for num in range(100, 200):
    if num % 37 == 0:
        num_new_1 = num % 100 *10 + num // 100
        print(num_new_1 , num % 100 *10, num // 100 )
        num_new_2 = num % 10 * 100 + num // 10
        print(num_new_2, num % 10 * 100 , num // 10)
        if num_new_1 % 37 != 0 or num_new_1 % 37 != 0:
            print("It's a false proposition.")
            break
else:
    print("It's a true proposition.")

111 110 1
111 100 11
481 480 1
814 800 14
851 850 1
518 500 18
It's a true proposition.


If a number is equal to the sum of its factors, the number is called the perfect number, for example, 6, because 6=1+2+3. Program and print all the perfect numbers within 1000.

In [22]:
#-*-coding:utf-8-*-
for i in range(1, 1001):
   s = 0   # pay attention to the position 
   for j in range(1, i):
       if i%j == 0:
            s += j
   if s == i:
       print("\n", i, " ", end = "")
       print("its factors are ", end = "")
       for j in range(1, i):
          if i%j == 0:
              print(j, end = " ")


 6  its factors are 1 2 3 
 28  its factors are 1 2 4 7 14 
 496  its factors are 1 2 4 8 16 31 62 124 248 

8. Verify one of the Goldbach Conjecture: the positive integer number (greater than or equal to 4) within 2000 can be decomposed into the sum of two prime numbers. Every even number's expression forming is such as: 4=2+2, print 6 expressions per line.

In [ ]:
import math
 
def prime(x):
    if x == 1:
        return False
    n = int(math.sqrt(x))
    for i in range(2, n+1):
        if x % i == 0:
            return False
    return True
 
def GC(n):
    k = 3      
    while k < n:
         t = n - k
         if t < k:
             break
         if prime(k) and prime(t):
             return k, t
         k += 2
 
n = int(input())
if n > 4:
    a,b = GC(n)
    print("{}={}+{}".format(n, a, b))
elif n == 4:
    print("{}={}+{}".format(4, 2, 2))

9. Write a program to let user input the number of apples and the unit price and then calculate the total price.
Use try-except statement to enable a program to handle non-numeric input. In case of non-numeric input, your program shall print error message and allow user to input again until a valid number. The program shall calculate the result upon a valid input is provided. The following is the expected result of your program.

In [25]:
while True:
    try:
        count = int(input("Enter count: "))
        price = float(input("Enter price for each one: "))
        Pay = count * price
        print("The price is: ", Pay)
        break
    except ValueError:
        print('Error, please enter numeric one. ')

The price is:  25.0
